lib/elixir/lib/map_set.ex

```diff
@@ -84,13 +94,40 @@ defmodule MapSet do
 
   """
   @spec difference(t, t) :: t
-  def difference(%MapSet{map: map1}, %MapSet{map: map2}) do
-    map = :maps.fold(fn value, _, acc ->
-      Map.delete(acc, value)
-    end, map1, map2)
+  # If the first set is less than twice the size of the second map,
+  # it is fastest to re-accumulate items in the first set that are not
+  # present in the second set.
+  def difference(%MapSet{map: map1}, %MapSet{map: map2})
+    when map_size(map1) < map_size(map2) * 2 do
+
+    map = map1
+    |> Map.keys
+    |> filter_not_in(map2)
+
     %MapSet{map: map}
   end
 
+  # If the second set is less than half the size of the first set, it's fastest
+  # to simply iterate through each item in the second set, deleting them from
+  # the first set.
+  def difference(%MapSet{map: map1}, %MapSet{map: map2}) do
+    %MapSet{map: Map.drop(map1, Map.keys(map2))}
+  end
+
+  def filter_not_in(keys, map2) do
```

Comment:
Indeed, also it'd better to define it as bodiless clause `def filter_not_in(keys, map2, acc \\ [])`.